### PR TITLE
sql: add missing locking for txnState.mu.txn writes

### LINKED ISSUE #32523: sql: TestIdleCancelSession failed under stress
SHA: https://github.com/cockroachdb/cockroach/commits/cfd53f62f038a06e34570f11ec02bc74dabd2aab

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stressrace TESTS=TestIdleCancelSession PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1021766&tab=buildLog

```
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1249 +0x1364
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:680 +0xac

Goroutine 589 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:202 +0x56d
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).execInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:499 +0x766
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).queryInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:249 +0x141
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).Query()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:233 +0x10e
  github.com/cockroachdb/cockroach/pkg/server.(*Server).clusterVersion()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server_update.go:162 +0xcd
  github.com/cockroachdb/cockroach/pkg/server.(*Server).upgradeStatus()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server_update.go:106 +0x8f
  github.com/cockroachdb/cockroach/pkg/server.(*Server).startAttemptUpgrade.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server_update.go:64 +0x27b
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:324 +0xf3

Goroutine 1273 (running) created at:
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:678 +0xb8
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport.(*http2Server).operateHeaders()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport/http2_server.go:417 +0x1446
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport.(*http2Server).HandleStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport/http2_server.go:457 +0x87a
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:676 +0x16e
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn.func2()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:631 +0x53
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn.func3()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:639 +0x34
==================
I181121 08:04:55.513024 89809 server/server_update.go:68  [n2] no need to upgrade, cluster already at the newest version
I181121 08:04:55.524198 90052 sql/event_log.go:126  [n2] Event: "node_join", target: 2, info: {Descriptor:{NodeID:2 Address:{NetworkField:tcp AddressField:127.0.0.1:36951} Attrs: Locality: ServerVersion:2.1-3 BuildTag:v2.2.0-alpha.20181119-229-gcfd53f6 StartedAt:1542787495052461629 LocalityAddress:[]} ClusterID:b9af0aae-d94e-4be7-b475-15ad76827f16 StartedAt:1542787495052461629 LastUp:1542787495052461629}
I181121 08:04:55.553948 90105 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181121 08:04:55.554007 90106 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181121 08:04:55.587560 90105 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
I181121 08:04:55.656168 89232 kv/transport_race.go:113  transport race promotion: ran 22 iterations on up to 301 requests
--- FAIL: TestIdleCancelSession (2.36s)
	testing.go:730: race detected during execution of test

```

### PR DESCRIPTION
**Pull request by petermattis**
Fixes #32523

Release note: None

### CODE REVIEW

**Comment by petermattis** [comment]
Note that I'm still trying to reproduce the original failure, though I'm pretty sure this is the fix.

File: pkg/sql/txn_state.go [hunk 1]
```diff
@@ -45,9 +45,11 @@ type txnState struct {
 	// Mutable fields accessed from goroutines not synchronized by this txn's
 	// session, such as when a SHOW SESSIONS statement is executed on another
 	// session.
+	//
 	// Note that reads of mu.txn from the session's main goroutine do not require
 	// acquiring a read lock - since only that goroutine will ever write to
-	// mu.txn.
+	// mu.txn. Writes to mu.txn do require a write lock to guarantee safety with
```
Comment (andreimatei) [inline pkg/sql/txn_state.go]: I'd argue that this comment is not needed, but I guess I've forfeited that right.

**Review by andreimatei — APPROVED** [review]
LGTM

**Comment by petermattis** [comment]
bors r=andreimatei

**Pull request merged**